pkg/pgmodel/model/pgsafetype/custom_pgtypes.go

```diff
@@ -0,0 +1,150 @@
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package pgsafetype
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgtype"
+)
+
+const (
+	NullChar         = '\x00'
+	NullCharSanitize = '\ufffe'
+)
+
+// Text is a custom pgtype that wraps the pgtype.Text. It is safe from null characters.
+type Text struct {
+	pgtype.Text
+}
+
+func (t *Text) DecodeBinary(ci *pgtype.ConnInfo, src []byte) error {
+	if err := t.Text.DecodeBinary(ci, src); err != nil {
+		return fmt.Errorf("pgsafetype.Text.DecodeBinary: %w", err)
+	}
+	return nil
+}
+
+func (t *Text) Set(src interface{}) error {
+	text, ok := src.(string)
+	if !ok {
+		return fmt.Errorf("pgsafetype.Text.Set(): src is not of 'string' type")
+	}
+	if err := t.Text.Set(sanitizeNullChars(text)); err != nil {
+		return fmt.Errorf("safe text: %w", err)
+	}
+	return nil
+}
+
+func (t *Text) Get() interface{} {
+	// Status is set to present in pgtype.Text when using pgtype.Text.Set(). However, when using a Scan(),
+	// this does not happen. In order to make this work, we need to explicitly tell here that data is present, otherwise
+	// data is not returned, rather a status is returned in pgtype.Text.Get().
+	t.Text.Status = pgtype.Present
+	s, ok := t.Text.Get().(string)
+	if !ok {
+		panic("'string' type not received from underlying 'pgtype.Text'")
+	}
+	return revertSanitization(s)
+}
+
+func (t *Text) AssignTo(_ interface{}) error {
+	panic("pgsafetype.Text.AssignTo(): not implemented")
+}
+
+func (t *Text) Scan(_ interface{}) error {
+	panic("pgsafetype.Text.Scan(): not implemented")
+}
+
+// TextArray is a custom pgtype that wraps the pgtype.TextArray. It is safe from null characters.
+type TextArray struct {
+	pgtype.TextArray
+}
+
+func (t *TextArray) DecodeBinary(ci *pgtype.ConnInfo, src []byte) error {
+	if err := t.TextArray.DecodeBinary(ci, src); err != nil {
+		return fmt.Errorf("pgsafetype.TextArray.DecodeBinary: %w", err)
+	}
+	return nil
+}
+
+func (t *TextArray) Set(src interface{}) error {
+	textArray, ok := src.([]string)
+	if !ok {
+		return fmt.Errorf("pgsafetype.TextArray.Set(): src is not of '[]string' type")
+	}
+	buf := make([]pgtype.Text, len(textArray))
+	for i := range textArray {
+		var text pgtype.Text
+		if err := text.Set(sanitizeNullChars(textArray[i])); err != nil {
+			return fmt.Errorf("safe text-array: %w", err)
+		}
+		buf[i] = text
+	}
+	if err := t.TextArray.Set(buf); err != nil {
+		return fmt.Errorf("safe textarray: %w", err)
+	}
+	return nil
+}
+
+func (t *TextArray) revertSanitization() {
+	val := t.TextArray.Get()
+	if val == nil {
+		return
+	}
+	arr := val.(pgtype.TextArray)
+	for i := range arr.Elements {
+		tmp, ok := arr.Elements[i].Get().(string)
+		if !ok {
+			panic("'string' type not received from underlying 'pgtype.Text' in 'pgsafetype.TextArray'")
+		}
+		arr.Elements[i].String = revertSanitization(tmp)
+	}
+	t.Elements = arr.Elements
```

Comment:
can't say I love this. if t.Elements is always sanitized that makes the logic easier to follow. Right now it can be sanitizes or unsanitized. This would lead to weird errors if we ever called TextArray.Get() and then sent the textarray as a parameter. Can we have this function return a new []string? This would prevent confusion []string is always real utf-8. pgutf8str.TextArrray is always sanitized

Reply:
Sorry, I cannot understand completely. You will only use a Get(), if the value in the type is already contained in the type, either by Set() or Scan()

Reply:
two changes
1) do not modify internal state during Get()
2. return []string and not TextArray from Get()

Reply:
Done, resolving.